File: ProblemSolve/203.cpp
/*******************************
 * Remove Linked List Elements *
 *******************************/

class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {        
        if(head == NULL){
            return NULL;
        }
        
        //현재 값이 제거해야 되는 값이면 미리 제거한다
        while(head!=NULL && head->val==val){
           head=head->next;
        }
        
        //현재 값이 아니라 다음 값부터 확인하면 된다
        ListNode* temp = head;
        
        //다음과 현재가 널이 아니면
        while(temp != NULL && temp -> next != NULL){
            //다음 값이 제거해야 하는 값이면 다음을 다음 다음으로 넘겨서 제거한다
            if(temp -> next -> val == val){
                temp -> next = temp -> next -> next;
                continue;
            }
            temp = temp -> next;
        }
        
        return head;
    }
};

File: ProblemSolve/203.cs
/*******************************
 * Remove Linked List Elements *
 *******************************/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
    public ListNode RemoveElements(ListNode head, int val) {
        if(head == null){
            return head;
        }
        
        //시작 값이 제거해야 하는 값이면 미리 제거
        while(head != null && head.val == val){
            head = head.next;
        }
        
        ListNode headTemp = head;
        
        while(headTemp != null && headTemp.next != null){
            //다음 노드를 지워야 하면 현재 노드의 다음 노드를 다다음 노드로 연결해버리면
            //다음 노드를 지워야 하는 것과 같은 효과를 낸다.
            if(headTemp.next.val == val){
                headTemp.next = headTemp.next.next;
                continue;
            }
            
            headTemp = headTemp.next;
        }
        
        return head;
    }
}